### ISSUE TITLE
Restrict Usernames to Plain ASCII

### ISSUE THREAD

**Issue by jwacalex** [engelsystem/engelsystem#606]
To prevent issues with unicode escaping or undefinable UTF-8 based nicknames, registration should only allow usernames match ``/^\w+$/``

**Comment by herzmut** [comment]
I have mixed feelings about excluding names like céline, yıldırım, tannhäuser, any name with a character like underscore _, period . or dashes - as well as all of the emojis, and I suggest to deal with Unicode inputs properly.

**Comment by jwacalex** [comment]
@herzmut the main issue is the "frontend" that has the issues. without a proper ascii representation it is currently not possible to find the angels because the names are at some random place in the lists.

**Comment by MyIgel** [comment]
The inputs are not the problem ;) The issue is that when someone wants to search a username it can be complicated "the a with a round on it and then the l with a cross".

Imho not allowing them isn't the right way, we should find an option to maybe convert them on the fly to ascii only and use that as an additional search field.

**Comment by herzmut** [comment]
On second degree search results, nicknames which are sorted with ascending Levenshtein distance could show up as suggestions. https://www.php.net/manual/en/function.levenshtein.php

**Comment by jwacalex** [comment]
Regarding levenshtein i want to quote the following remark 

>  The levenshtein function processes each byte of the input string individually. Then for multibyte encodings, such as UTF-8, it may give misleading results.

https://www.php.net/manual/en/function.levenshtein.php#113702

**Comment by herzmut** [comment]
Despite the fact that there are multibyte-compatible implementations, I quickly tested that proposition with the names mentioned above and found the results to be negligible: For the names mentioned above, each unicode character counted as 2 instead of 1 in their Levensthein distance result compared to their respective ASCII equivalent. However, another obstacle of Levensthein in PHP is the limitation on 255 characters.

**Comment by MyIgel** [comment]
That only gets more complicated when using "non latin only" characters, the comparison function should be aware of it. But that needs some further investigation imho and maybe a test implementation comparing the various possibilities / methods

**Comment by teal-bauer** [comment]
There are various foldings and normalized forms for Unicode strings, it wouldn't be a ton of work to make e.g. an input like `celine` match `céline`, or `yildirim` match `yıldırım`. It might not be possible to do on the fly in the DB, so adding a separate column for "normalized" nickname or a separate column only for searching containing stemmed and normalized forms of all searchable fields might be required / useful.

**Comment by teal-bauer** [comment]
Looked into this - the admin search pages (free angels, arrivals, active users) currently load all matching users into memory and filter in PHP rather than querying the DB. Will open a simple fix for Unicode-aware filtering and a follow-up issue about improving the search architecture.

**Comment by teal-bauer** [comment]
PR #1639 already touches the same code for related issue #1317 (case-insensitive umlaut matching). Will extend that branch with ASCII folding.

**Comment by teal-bauer** [comment]
And I guess no follow-up issue is needed, even the largest events will not have such a high amount of angels that the memory search won't scale I suppose…

**Comment by MyIgel** [comment]
That was the idea behind the "jeah just dump it all"; when we have so big events that it matters imho a dedicated search instance (as in: external service) would be the better solution (and could be used for fuzzy searching "everywhere" too at some point)

**Comment by teal-bauer** [comment]
Yeah, so I would say that #1639 closes this even if in a different way :)
